src/canvas-force-graph.js

```diff
@@ -154,60 +178,72 @@ export default Kapsule({
 
         const visibleLinks = state.graphData.links.filter(getVisibility);
 
-        if (state.linkCanvasObject) {
-          // Custom link paints
-          visibleLinks.forEach(link => state.linkCanvasObject(link, state.ctx, state.globalScale));
-          return;
-        }
-
         // Bundle strokes per unique color/width for performance optimization
-        const linksPerColor = indexBy(visibleLinks, [getColor, getWidth]);
+        const linksPerColor = indexBy(visibleLinks, [getColor, getWidth, getLinkCanvasObjectMode]);
 
         Object.entries(linksPerColor).forEach(([color, linksPerWidth]) => {
           const lineColor = !color || color === 'undefined' ? 'rgba(0,0,0,0.15)' : color;
-          Object.entries(linksPerWidth).forEach(([width, links]) => {
+          Object.entries(linksPerWidth).forEach(([width, linksPerCanvasObjectMode]) => {
             const lineWidth = (width || 1) / state.globalScale + padAmount;
-
-            ctx.beginPath();
-            links.forEach(link => {
-              const start = link.source;
-              const end = link.target;
-              if (!start.hasOwnProperty('x') || !end.hasOwnProperty('x')) return; // skip invalid link
-
-              const curvature = getCurvature(link);
-
-              ctx.moveTo(start.x, start.y);
-
-              if (!curvature) { // Straight line
-                ctx.lineTo(end.x, end.y);
-                link.__controlPoints = null;
-                return;
+            Object.entries(linksPerCanvasObjectMode).forEach(([linkCanvasObjectMode, links]) => {
```

Comment:
This iteration over the modes seems too coupled with the link default drawing. I suggest that we run the `linkCanvasObjectObjectMode` accessor once per link, just like in the nodes and keep only one level of iteration.